File: packages/sdk/src/conversation/invokeResponseFactory.ts
import { InvokeResponse, StatusCodes } from "botbuilder";
import { InvokeResponseErrorCode } from "./interface";

/**
 * Available response type for an adaptive card invoke response.
 * @internal
 */
export enum InvokeResponseType {
  AdaptiveCard = "application/vnd.microsoft.card.adaptive",
  Message = "application/vnd.microsoft.activity.message",
  Error = "application/vnd.microsoft.error",
}

/**
 * Provides methods for formatting various invoke responses a bot can send to respond to an invoke request.
 *
 * @remarks
 * All of these functions return an {@link InvokeResponse} object, which can be
 * passed as input to generate a new `invokeResponse` activity.
 *
 * This example sends an invoke response that contains an adaptive card.
 *
 * ```typescript
 *
 * const myCard = {
 *    type: "AdaptiveCard",
 *    body: [
 *     {
 *       "type": "TextBlock",
 *       "text": "This is a sample card"
 *     }],
 *     $schema: "http://adaptivecards.io/schemas/adaptive-card.json",
 *     version: "1.4"
 *  };
 *
 * const invokeResponse = InvokeResponseFactory.adaptiveCard(myCard);
 * await context.sendActivity({
 *    type: ActivityTypes.InvokeResponse,
 *    value: invokeResponse,
 *  });
 * ```
 */
export class InvokeResponseFactory {
  /**
   * Create an invoke response from a text message.
   * The type of the invoke response is `application/vnd.microsoft.activity.message`
   * indicates the request was successfully processed.
   *
   * @param message - A text message included in a invoke response.
   *
   * @returns An `InvokeResponse` object.
   */
  public static textMessage(message: string): InvokeResponse {
    if (!message) {
      throw new Error("The text message cannot be null or empty");
    }

    return {
      status: StatusCodes.OK,
      body: {
        statusCode: StatusCodes.OK,
        type: InvokeResponseType.Message,
        value: message,
      },
    };
  }

  /**
   * Create an invoke response from an adaptive card.
   *
   * The type of the invoke response is `application/vnd.microsoft.card.adaptive` indicates
   * the request was successfully processed, and the response includes an adaptive card
   * that the client should display in place of the current one.
   *
   * @param card - The adaptive card JSON payload.
   *
   * @returns An `InvokeResponse` object.
   */
  public static adaptiveCard(card: unknown): InvokeResponse {
    if (!card) {
      throw new Error("The adaptive card content cannot be null or undefined");
    }

    return {
      status: StatusCodes.OK,
      body: {
        statusCode: StatusCodes.OK,
        type: InvokeResponseType.AdaptiveCard,
        value: card,
      },
    };
  }

  /**
   * Create an invoke response with error code and message.
   *
   * The type of the invoke response is `application/vnd.microsoft.error` indicates
   * the request was failed to processed.
   *
   * @param errorCode - The status code indicates error, available values:
   *  - 400 (BadRequest): indicate the incoming request was invalid.
   *  - 500 (InternalServerError): indicate an unexpected error occurred.
   * @param errorMessage - The error message.
   *
   * @returns An `InvokeResponse` object.
   */
  public static errorResponse(
    errorCode: InvokeResponseErrorCode,
    errorMessage: string
  ): InvokeResponse {
    return {
      status: StatusCodes.OK,
      body: {
        statusCode: errorCode,
        type: InvokeResponseType.Error,
        value: {
          code: errorCode.toString(),
          message: errorMessage,
        },
      },
    };
  }

  /**
   * Create an invoke response with status code and response value.
   * @param statusCode - The status code.
   * @param body - The value of the response body.
   *
   * @returns An `InvokeResponse` object.
   */
  public static createInvokeResponse(statusCode: StatusCodes, body?: unknown): InvokeResponse {
    return {
      status: statusCode,
      body: body,
    };
  }
}

File: packages/dotnet-sdk/src/TeamsFx/Conversation/InvokeResponseFactory.cs
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

namespace Microsoft.TeamsFx.Conversation
{
    using Microsoft.Bot.Builder;
    using Microsoft.Bot.Schema;
    using System.Net;

    /// <summary>
    /// Contains utility methods for various invoke response types for an adaptiveCard/action invoke response.
    /// </summary>
    public static class InvokeResponseFactory
    {
        /// <summary>
        /// Creates invoke response with an adaptive card.
        /// </summary>
        /// <param name="card">The adaptive card included in the response body.</param>
        /// <returns>An instance of <see cref="InvokeResponse"/></returns>
        public static InvokeResponse AdaptiveCard(object card)
        {
            var response = new AdaptiveCardInvokeResponse()
            {
                StatusCode = 200,
                Type = InvokeResponseContentType.AdaptiveCard,
                Value = card
            };

            return new InvokeResponse
            {
                Status = (int)HttpStatusCode.OK,
                Body = response
            };
        }

        /// <summary>
        /// Creates invoke response with a text message.
        /// </summary>
        /// <param name="message">The text message included in the response body.</param>
        /// <returns>An instance of <see cref="InvokeResponse"/>.</returns>
        public static InvokeResponse TextMessage(string message)
        {
            if (string.IsNullOrWhiteSpace(message))
            {
                throw new ArgumentNullException(nameof(message));
            }

            var response = new AdaptiveCardInvokeResponse()
            {
                StatusCode = 200,
                Type = InvokeResponseContentType.Message,
                Value = message
            };

            return new InvokeResponse
            {
                Status = (int)HttpStatusCode.OK,
                Body = response
            };
        }

        /// <summary>
        /// Creates invoke response with error code and error message.
        /// </summary>
        /// <param name="errorCode">The status code indicates the bot processing error, available values:
        /// <list type="bullet">
        ///     <item>
        ///         <description> 400 (BadRequest): indicate the incoming request was invalid.</description>
        ///     </item>
        ///     <item>
        ///         <description> 500 (InternalServerError): indicate an unexpected error occurred.</description>
        ///     </item>
        /// </list>
        /// </param>
        /// <param name="errorMessage">The error message.</param>
        /// <returns>An instance of <see cref="InvokeResponse"/></returns>
        public static InvokeResponse ErrorResponse(InvokeResponseErrorCode errorCode, string errorMessage)
        {
            if (string.IsNullOrWhiteSpace(errorMessage))
            {
                throw new ArgumentNullException(nameof(errorMessage));
            }

            var response = new AdaptiveCardInvokeResponse()
            {
                StatusCode = (int)errorCode,
                Type = InvokeResponseContentType.Error,
                Value = new Error()
                {
                    Code = errorCode.ToString(),
                    Message = errorMessage
                }
            };

            return new InvokeResponse
            {
                Status = (int)HttpStatusCode.OK,
                Body = response
            };
        }

        /// <summary>
        /// Creates invoke response with status code and response body.
        /// </summary>
        /// <param name="statusCode">The status code.</param>
        /// <param name="response">The response body.</param>
        /// <returns>An instance of <see cref="InvokeResponse"/>.</returns>
        public static InvokeResponse CreateInvokeResponse(HttpStatusCode statusCode, object response)
        {
            return new InvokeResponse
            {
                Status = (int)statusCode,
                Body = response
            };
        }
    }
}
